### PR TITLE
Ehrenfest dynamics

### PR DESCRIPTION
**Pull request by fbonafe**
This pull request that includes electronic and Ehrenfest dynamics, capabilities for pump-probe simulations, updated manual and regression tests.

### CODE REVIEW

File: prog/dftb+/lib_timedep/dynamics.F90 [hunk 1]
```diff
@@ -0,0 +1,3317 @@
...
+      qBlock(:,:,:,:) = 0.0_dp
+      do iKS = 1, this%parallelKS%nLocalKS
+        iK = this%parallelKS%localKS(1, iKS)
+        iSpin = this%parallelKS%localKS(2, iKS)
+        do iAt = 1, this%nAtom
+          iOrb1 = iSquare(iAt)
+          iOrb2 = iSquare(iAt+1)
+          nOrb = iOrb2 - iOrb1
+          qBlock(:nOrb,:nOrb,iAt,iSpin) = qBlock(:nOrb,:nOrb,iAt,iSpin) + this%kWeight(iK) *&
+              & real(matmul(Ssqr(iOrb1:iOrb2-1,:,iSpin), rho(:,iOrb1:iOrb2-1,iSpin)))
```
Comment (aradi) [inline prog/dftb+/lib_timedep/dynamics.F90]: NOTE: you loose the last 6 digits here (missing `kind=dp` in `real()`). Compare
```
  print *, 10.0_dp * real(0.1_dp)
  print *, 10.0_dp * real(0.1_dp, dp)
```

**Comment by aradi** [comment]
@fbonafe If you fix the issues mentioned above, we would be happy to merge this branch and close the PR. :wink:

**Comment by aradi** [comment]
@fbonafe I know, you are probably quite busy now. Would you have time to fix those few things I asked for, so that we can merge to master?

**Comment by aradi** [comment]
@fbonafe  And, please. bring it also up-to date with current master, so that we do not have merge conflicts. If you need help, let me know! :smile:

**Comment by bhourahine** [comment]
> Very nice PR, well written and very clear. I marked a few minor issues. And in a separate commit, could you reindent `dynamics.F90` (as the indendation is switching between 2 and 3 spaces) and break the few lines longer than 100 chars?
[3990bea3](https://github.com/dftbplus/dftbplus/pull/329/commits/3990bea3e6f6ee65495ff4fc6dfe5ba1e87a9ca2)

File: doc/dftb+/manual/dftbp.tex [hunk 2]
```diff
@@ -3268,6 +3269,265 @@ \subsubsection{Localise}
 
 \end{description}
 
+\subsection{ElectronDynamics}
+
```
[the remark this answers is not in the document]
Comment (bhourahine) [reply doc/dftb+/manual/dftbp.tex:3936]: Suggest a separate PR after the merge

File: doc/dftb+/manual/dftbp.tex [hunk 3]
```diff
@@ -3268,6 +3269,265 @@ \subsubsection{Localise}
...
+\begin{description}
+\item[\is{Steps}] Number of propagation steps to perform.
+
+\item[\is{TimeStep}]\modif{\modtype{time}} Time interval between two
+  electronic propagation steps.
+
+\item[\is{FieldStrength}]\modif{\modtype{Electric field strength}} Peak intensity of the applied time-dependent electric field.
+
+\item[\is{Populations}] If time-dependent (ground state) molecular orbital populations should be calculated. This is done by projecting the electronic state onto the ground state molecular orbitals. If Yes, they are saved every \is{WriteFrequency} steps to \verb|molPopul1.dat| (and \verb|molPopul2.dat| is the calculation is collinearly spin polarised).
```
[the remark this answers is not in the document]
Comment (bhourahine) [reply doc/dftb+/manual/dftbp.tex:3974]: Suggests the need of a PR for the other excited state routines using ALL CAPS for files. Doesn't match code behaviour anyway.

File: prog/dftb+/lib_timedep/dynamics.F90 [hunk 4]
```diff
@@ -0,0 +1,3317 @@
...
+
+    !> Reference external potential (usual provided via API)
+    type(TRefExtPot) :: refExtPot
+
+    complex(dp) :: Ssqr(this%nOrbs,this%nOrbs,this%parallelKS%nLocalKS)
+    complex(dp) :: Sinv(this%nOrbs,this%nOrbs,this%parallelKS%nLocalKS)
+    complex(dp), target :: trho(this%nOrbs,this%nOrbs,this%parallelKS%nLocalKS)
+    complex(dp), target :: trhoOld(this%nOrbs,this%nOrbs,this%parallelKS%nLocalKS)
+    complex(dp), pointer :: rho(:,:,:), rhoOld(:,:,:)
+    complex(dp) :: H1(this%nOrbs,this%nOrbs,this%parallelKS%nLocalKS)
```
[the remark this answers is not in the document]
Comment (bhourahine) [reply prog/dftb+/lib_timedep/dynamics.F90]: Various other large arrays like this. Switched to allocating explicitly.

File: utils/calc_timeprop_spectrum [hunk 5]
```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env python3
+
```
[the remark this answers is not in the document]
Comment (bhourahine) [reply utils/calc_timeprop_spectrum:2]: One for a later PR to get into dp_tools

File: prog/dftb+/lib_dftb/sparse2dense.F90 [hunk 6]
```diff
@@ -1124,6 +1125,74 @@ subroutine packHSPauliImag_kpts(primitive, square, kPoint, kWeight, iNeighbour,
   end subroutine packHSPauliImag_kpts
 
 
+  !> Pack squared matrix in the sparse form (real version).
+  subroutine packHS_hermitian(prim, iPrim, square, iNeighbour, nNeighbourSK, mOrb, iAtomStart,&
+      & iPair, img2CentCell)
```
[the remark this answers is not in the document]
Comment (bhourahine) [reply prog/dftb+/lib_dftb/sparse2dense.F90:1136]: Staying inside the same interface? Have renamed the internals.

File: doc/dftb+/manual/dftbp.tex [hunk 7]
```diff
@@ -3268,6 +3269,265 @@ \subsubsection{Localise}
...
+
+\item[\is{Probe}] If this is the simulation of a probe. If set to yes, then automatically \is{Restart} = Yes, \is{WriteRestart} = No, and only the dipole moment output files will be written, since this keyword is used together with a \is{Kick} perturbation to probe the system. Notice that the dump file that contains the state of the system that wants to be probed must be present and renamed to \verb|tddump.bin|.
+
+\item[\is{EulerFrequency}] Number of steps every which an Euler integration step is done in the electronic propagation (that normally uses the Leapfrog algorithm). The default value ensures no Euler steps are done during the dynamics. If used, must be set to something larger than 50.
+
+\item[\is{WriteBondEnergy}] If the pairwise, time-dependent bond energy $BE(t)$, calculated as
+$$ BE_{AB}(t) = \sum_{\mu \in A} \sum_{\nu \in B} \rho_{\mu \nu}(t) H^0_{\mu \nu}(t),$$
+should be written to the \verb|bondenergy.bin| file. This is written every \is{WriteFrequency} steps, and in binary format because of the large amount of data. Currently incompatible with \is{WriteBondOrder}.
```
[the remark this answers is not in the document]
Comment (fbonafe) [reply doc/dftb+/manual/dftbp.tex]: Basically we use the same array and subroutines inside the code to compute both, and I couldn't think of an easy and practical way to allow for both without duplicating code. Maybe adding an extra index to this array, similar to the spin index, which is either 1 or 2 depending on whether one quantity or both are requested?

File: doc/dftb+/manual/dftbp.tex (hunk 2 above)
[the remark this answers is not in the document]
Comment (fbonafe) [reply doc/dftb+/manual/dftbp.tex:3936]: I agree with doing it in a separate PR, and also agree that should be a top-level block.

**Comment by bhourahine** [comment]
Manually merged after a filter branch to remove some large binary files. Can definitely recommend [bfg](https://rtyley.github.io/bfg-repo-cleaner).

**Pull request closed without merging**